compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -693,6 +694,45 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         );
     }
 
+    pub(crate) fn note_result_coercion(
+        &self,
+        err: &mut Diagnostic,
+        expr: &hir::Expr<'tcx>,
+        expected: Ty<'tcx>,
+        found: Ty<'tcx>,
+    ) -> bool {
+        let ty::Adt(e, substs_e) = expected.kind() else { return false; };
+        let ty::Adt(f, substs_f) = found.kind() else { return false; };
+        if e.did() != f.did() {
+            return false;
+        }
+        if Some(e.did()) != self.tcx.get_diagnostic_item(sym::Result) {
+            return false;
+        }
+        let e = substs_e.type_at(1);
+        let f = substs_f.type_at(1);
+        if self
+            .infcx
+            .type_implements_trait(
+                self.tcx.get_diagnostic_item(sym::Into).unwrap(),
+                [self.tcx.erase_regions(f), self.tcx.erase_regions(e)],
+                self.param_env,
+            )
+            .must_apply_modulo_regions()
+        {
+            err.multipart_suggestion(
+                "you can rely on the implicit conversion that `?` does to transform the error type",
```

Comment:
This phrasing feels a bit roundabout... Also, it should really mention that it's doing control flow, since it's not equivalent to `.map_err(Into::into)`, which is what this currently reads like.

Reply:
In return position, it's fine to ignore the fact that this is doing control flow, but even if you hadn't applied my suggestion to extend this, it's still happening on any block-tail usage of `CoerceMany`, which is not always in the return of an fn.

Reply:
Before the requested change, this *was* checking the obligation cause code explicitly for the cases where this was happening in `return` statements and function tails exclusively :-/

Reply:
> where this was happening in return statements and function tails exclusively :-/

Wait, like I said in:

> but even if you hadn't applied my suggestion to extend this to all type mismatch suggestions, it's still happening on any block-tail usage of CoerceMany, which is not always in the return of an fn.

This concern still applies to the old code too. 

The original commit bb0f8c1098d7191f1c27ce7ee3ee2bafcf99c6ca used the `BlockTailExpression` obligation to gate this suggestion, which isn't equivalent to checking if something is in the function tail exclusively. For example:

```
struct A;
struct B;

impl From<A> for B {
    fn from(_: A) -> Self { B }
}

fn foo() -> Result<(), B> {
    let y = Err(A);
    let z: Result<(), B> = { y };
}
```

Gives:

```
help: you can rely on the implicit conversion that `?` does to transform the error type
   |
10 |     let z: Result<(), B> = { Ok(y?) };
   |                              +++ ++
```

On that earlier commit, which still really deserves either fixing (making the suggestion *only* apply in tail position), or mentioning control flow.

Reply:
Fixed

Reply:
Regarding the wording, do you have one in mind? I'm somewhat relying on prior understanding of what `?` does, but because the suggestion is always for returned values, it should never affect behavior, regardless of understanding.